devdoc/singlylinkedlist2_requirements.md

```diff
@@ -0,0 +1,176 @@
+typedef int (*SLIST_ACTION_FUNCTION)(PSLIST_ENTRY listEntry, void* action_context, bool* continue_processing);
+
+MOCKABLE_FUNCTION(, void, SList_InitializeListHead, PSLIST_ENTRY, listHead);
+MOCKABLE_FUNCTION(, int, SList_IsListEmpty, const PSLIST_ENTRY, listHead);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_Add, PSLIST_ENTRY, listHead, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_AddHead, PSLIST_ENTRY, listHead, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, int, SList_Remove, PSLIST_ENTRY, listHead, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_RemoveHead, PSLIST_ENTRY, listHead);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_GetNextItem, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_Find, PSLIST_ENTRY, listHead, SLIST_MATCH_FUNCTION, matchFunction, void*, matchContext);
+MOCKABLE_FUNCTION(, int, SList_RemoveIf, PSLIST_ENTRY, listHead, SLIST_CONDITION_FUNCTION, conditionFunction, void*, matchContext);
+MOCKABLE_FUNCTION(, int, SList_ForEach, PSLIST_ENTRY, listHead, SLIST_ACTION_FUNCTION, actionFunction, void*, actionContext);
+
+```
+## Usage
+
+`SLIST_ENTRY` does not store any data. Instead, `SLIST_ENTRY` is placed inside a struct:
+
+```c
+typedef struct DATA_TAG
+{
+    int num1;
+    int num2;
+    SLIST_ENTRY anchor;
+} DATA;
+```
+
+The parent struct of a list entry can be obtained using `CONTAINING_RECORD`:
+
+```
+PSLIST_ENTRY listEntry = SList_RemoveHead(listHead);
+DATA* data = CONTAINING_RECORD(listEntry, DATA, anchor);
+```
+
+### SList_InitializeListHead
+```c
+void SList_InitializeListHead(PSLIST_ENTRY listHead);
+```
+SList_InitializeListHead will initialize the listHead next pointer points to NULL. 
+
+### SList_IsListEmpty
+```c
+int SList_IsListEmpty(const PSLIST_ENTRY listHead);
+```
+
+SList_IsListEmpty shall return a non-zero value if there are no SLIST_ENTRY's on this list other than the list head. 
+
+SList_IsListEmpty shall return 0 if there is one or more items in the list. 
+
+Notes:
+1.	SList_IsListEmpty shall be undefined if the SLIST_ENTRY is not currently part of a list.
+2.	SList_IsListEmpty shall be undefined if the SLIST_ENTRY has not been initialized as with SList_InitializeListHead.
+
+### SList_Add
+```c
+PSLIST_ENTRY SList_Add(PSLIST_ENTRY listHead, PSLIST_ENTRY listEntry);
+```
+
+SList_Add shall add one item to the tail of the list and on success it shall return an entry to the added item. 
+
+If any of the arguments is NULL, SList_Add shall not add the item to the list and return NULL. 
+
+If allocating the new list node fails, SList_Add shall return NULL. 
+
+### SList_AddHead
+```
+PSLIST_ENTRY SList_AddHead(PSLIST_ENTRY listHead, PSLIST_ENTRY listEntry);
+```
+
+`SList_AddHead` shall insert `item` at head on success it shall return an entry to the added item. 
+
+If any of the arguments is NULL, SList_AddHead shall not add the item to the list and return NULL. 
```

Comment:
>If any of the arguments is NULL, SList_AddHead shall not add the item to the list and return NULL. [](http://example.com/codeflow?start=0&length=98)

Can you list out the arguments separately so you can write a test for each of them.  For example this would say

If `listHead` is `NULL`, `SList_AddHead` shall not add the item to the list and return `NULL`

and another one for listEntry

Reply:
done